velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -62,57 +66,415 @@ void generateJsonTyped(
+      for (int i = offset + 1; i < offset + size; ++i) {
+        proxy.append(","_sv);
+        proxy.append(elementsStringResult->valueAt(elementsDecoded->index(i)));
+      }
+      proxy.append("]"_sv);
+
+      proxy.finalize();
+    }
+  });
+}
+
+FOLLY_ALWAYS_INLINE void appendPair(
+    const FlatVector<StringView>& keysStrings,
+    const FlatVector<StringView>& valuesStrings,
+    size_t keyOffset,
+    size_t valueOffset,
+    exec::StringProxy<FlatVector<StringView>>& out) {
+  out.append(keysStrings.valueAt(keyOffset));
+  out.append(":"_sv);
+  out.append(valuesStrings.valueAt(valueOffset));
+}
+
+void castToJsonFromMap(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult) {
+  // input is guaranteed to be in flat encoding when passed in.
+  auto inputMap = input.as<MapVector>();
+
+  auto mapKeys = inputMap->mapKeys();
+  VELOX_CHECK_EQ(
+      mapKeys->typeKind(),
+      TypeKind::VARCHAR,
+      "The type of map keys must be VARCHAR.");
+
+  auto elementsRows = functions::toElementRows(mapKeys->size(), rows, inputMap);
+  exec::LocalDecodedVector keysDecoded(context, *mapKeys, elementsRows);
+
+  exec::LocalSelectivityVector keysBaseRows(
+      context->execCtx(), keysDecoded->base()->size());
+  translateRowsToBase(*keysDecoded, elementsRows, *keysBaseRows);
+
+  auto mapValues = inputMap->mapValues();
+  exec::LocalDecodedVector valuesDecoded(context, *mapValues, elementsRows);
+
+  exec::LocalSelectivityVector valuesBaseRows(
+      context->execCtx(), valuesDecoded->base()->size());
+  translateRowsToBase(*valuesDecoded, elementsRows, *valuesBaseRows);
+
+  VectorPtr keysResult;
+  BaseVector::ensureWritable(
+      *keysBaseRows, JSON(), context->pool(), &keysResult);
+  auto keysStringResult = keysResult->as<FlatVector<StringView>>();
+
+  // The caller of castTo() should have ensured map keys are VARCHAR. In
+  // addition, map keys must not be null.
+  castToJson<TypeKind::VARCHAR, true>(
+      *keysDecoded->base(), context, *keysBaseRows, *keysStringResult);
+
+  VectorPtr valuesResult;
+  BaseVector::ensureWritable(
+      *valuesBaseRows, JSON(), context->pool(), &valuesResult);
+  auto valuesStringResult = valuesResult->as<FlatVector<StringView>>();
+
+  VELOX_DYNAMIC_TYPE_DISPATCH(
+      castToJson,
+      mapValues->typeKind(),
+      *valuesDecoded->base(),
+      context,
+      *valuesBaseRows,
+      *valuesStringResult);
+
+  // Reserves space for all keys and values Json strings and the delimiters to
+  // be added.
+  auto elementsStringSize = estimateJsonStringSizeForArraysAndMaps(
+      *keysStringResult,
+      keysDecoded->indices(),
+      inputMap->rawOffsets(),
+      inputMap->rawSizes(),
+      rows);
+  elementsStringSize += estimateJsonStringSizeForArraysAndMaps(
+      *valuesStringResult,
+      valuesDecoded->indices(),
+      inputMap->rawOffsets(),
+      inputMap->rawSizes(),
+      rows);
```

Comment:
I'm not sure these functions are really helping; just writing out the loop here is more concise, clearer, and faster (since you only iterate over the arrays once). Assuming you're using that helper struct, and with fixes for nulls:
```suggestion
  size_t elementsStringSize = 0;
  rows.applyToSelected([&](auto row) {
    if (inputMap->isNullAt(i)) {
      elementsStringSize += 4;
      return;
    }
    auto offset = inputMap->offsetAt(row);
    for (auto i = offset, end = offset + inputMap->sizeAt(row); i != end; ++i) {
      elementsStringSize += keys.at(i).size() + values.at(i).size();
    }
  }
```